x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,295 @@
+
+    private static final ParseField RULESET_ID_FIELD = new ParseField("ruleset_id");
+    private static final ParseField MATCH_CRITERIA_FIELD = new ParseField("match_criteria");
+    private static final ParseField ORGANIC_QUERY_FIELD = new ParseField("organic");
+
+    /**
+     * Defines the set of allowed match criteria, so that we can validate that rule query requests are sending in allowed/supported data.
+     */
+    static final Set<String> ALLOWED_MATCH_CRITERIA = Set.of("query_string");
+
+    private final String rulesetId;
+    private final Map<String, Object> matchCriteria;
+    private final QueryBuilder organicQuery;
+
+    private final List<String> pinnedIds;
+    private final Supplier<List<String>> pinnedIdsSupplier;
+    private final List<Item> pinnedDocs;
+    private final Supplier<List<Item>> pinnedDocsSupplier;
+
+    private final Logger logger = LogManager.getLogger(RuleQueryBuilder.class);
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_500_024;
+    }
+
+    public RuleQueryBuilder(QueryBuilder organicQuery, Map<String, Object> matchCriteria, String rulesetId) {
+        this(organicQuery, matchCriteria, rulesetId, null, null, null, null);
+    }
+
+    public RuleQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        organicQuery = in.readNamedWriteable(QueryBuilder.class);
+        matchCriteria = in.readMap();
+        rulesetId = in.readString();
+        pinnedIds = in.readOptionalStringList();
+        pinnedIdsSupplier = null;
+        pinnedDocs = in.readOptionalList(Item::new);
+        pinnedDocsSupplier = null;
+    }
+
+    private RuleQueryBuilder(
+        QueryBuilder organicQuery,
+        Map<String, Object> matchCriteria,
+        String rulesetId,
+        List<String> pinnedIds,
+        List<Item> pinnedDocs,
+        Supplier<List<String>> pinnedIdsSupplier,
+        Supplier<List<Item>> pinnedDocsSupplier
+
+    ) {
+        if (organicQuery == null) {
+            throw new IllegalArgumentException("organicQuery must not be null");
+        }
+        if (matchCriteria == null || matchCriteria.isEmpty()) {
+            throw new IllegalArgumentException("matchCriteria must not be null or empty");
+        }
+        for (String matchCriteriaKey : matchCriteria.keySet()) {
+            if (ALLOWED_MATCH_CRITERIA.contains(matchCriteriaKey) == false) {
+                throw new IllegalArgumentException("matchCriteria key [" + matchCriteriaKey + "] is not allowed");
+            }
+        }
+        if (Strings.isNullOrEmpty(rulesetId)) {
+            throw new IllegalArgumentException("rulesetId must not be null or empty");
+        }
+
+        // PinnedQueryBuilder will return an error if we attmept to return more than the maximum number of
+        // pinned hits. Here, we truncate matching rules rather than return an error.
+        if (pinnedIds != null && pinnedIds.size() > MAX_NUM_PINNED_HITS) {
```

Comment:
Why not throwing an error? Truncating silently would be confusing?

Reply:
I have a followup PR that will return a header warning: https://github.com/elastic/elasticsearch/pull/97473